link/link.go

```diff
@@ -446,3 +461,74 @@ func (l *RawLink) Info() (*Info, error) {
 		extra,
 	}, nil
 }
+
+// Iterator allows iterating over links attached into the kernel.
+type Iterator struct {
+	// The ID of the current link. Only valid after a call to Next
+	ID ID
+	// The current link. Only valid until a call to Next.
+	// See Take if you want to retain the link.
+	Link Link
+	err  error
+}
+
+// Next retrieves the next link.
+//
+// Returns true if another link was found. Call [Iterator.Err] after the function returns false.
+func (it *Iterator) Next() bool {
+	id := it.ID
+	for {
+		getIdAttr := &sys.LinkGetNextIdAttr{Id: id}
+		err := sys.LinkGetNextId(getIdAttr)
+		if errors.Is(err, os.ErrNotExist) {
+			// There are no more links.
+			break
+		} else if err != nil {
+			it.err = fmt.Errorf("get next link ID: %w", err)
+			break
+		}
+
+		id = getIdAttr.NextId
+		l, err := NewLinkFromID(id)
+		if errors.Is(err, os.ErrNotExist) {
+			// Couldn't load the link fast enough. Try next ID.
+			continue
+		} else if err != nil {
+			it.err = fmt.Errorf("get link for ID %d: %w", id, err)
+			break
+		}
+
+		if it.Link != nil {
```

Comment:
the iterator code lgtm, but I don't understand why are links being closed? wouldn't this break existing links?

Reply:
Good point. The link lifecycle is not super clear to me. If you have resources on this they are highly appreciated.
From what I can gather from [here](https://kernel.org/doc/html/v6.9-rc1/userspace-api/ebpf/syscall.html) and [here](https://ebpf-go.dev/concepts/object-lifecycle/). 

> BPF_LINK_GET_FD_BY_ID
...
Return
A new file descriptor (a nonnegative integer), or -1 if an error occurred (in which case, errno is set appropriately).

and at the end (although this refers to maps and programs, not specifically to links)
> An eBPF object is deallocated only after all file descriptors referring to the object have been closed and no references remain pinned to the filesystem or attached (for example, bound to a program or device).

So my rough understanding is that only the new file descriptor is closed on `Close()`. If there are still references from other sources the link will remain active.

I wonder if there is an easy way to validate it in unit tests. 
Calling `testLink(t, &linkCgroup{*tLink}, prog)` on the original (Raw) link, after the iterator is done seems to work.
Not sure if this is enough though...

Reply:
> An eBPF object is deallocated only after all file descriptors referring to the object have been closed and no references remain pinned to the filesystem or attached

ok, I see 👍🏻